enhancements/20240611-vm-schedule-backup.md

```diff
@@ -0,0 +1,171 @@
+
+## Proposal
+
+### User Stories
+
+Although [Longhorn provides recurring backup](https://longhorn.io/docs/1.6.2/snapshots-and-backups/scheduling-backups-and-snapshots/) that is per volume and not VM-based, so it brings the challenge to restore the VM from these volume backups.
+
+With this feature, the user can have a **list of VM backups** which are kept renovated according to the format of a [CRON expression](https://en.wikipedia.org/wiki/Cron#CRON_expression), and the user also can choose anyone from the list to restore the VM.
+
+
+### API changes
+
+## Design
+
+### Implementation Overview
+
+#### CRD
+
+- Add new resource `ScheduleVMBackup` to control VM schedule backup jobs
+```
+type ScheduleVMBackupSpec struct {
+	// +kubebuilder:validation:Required
+	Cron string `json:"cron"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=8
+	// +kubebuilder:validation:Maximum=254
+	// +kubebuilder:validation:Minimum=2
+	Retain int `json:"retain"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default:=4
+	// +kubebuilder:validation:Minimum=2
+	MaxFailure int `json:"maxFailure"`
+
+	// +optional
+	ResumeRequest bool `json:"resumeRequest"`
+
+	// +kubebuilder:validation:Required
+	VMBackupSpec VirtualMachineBackupSpec `json:"vmbackup"`
+}
+```
+
+```
+type ScheduleVMBackupStatus struct {
+	// +optional
+	VMBackupInfo []VMBackupInfo `json:"vmbackupInfo,omitempty"`
+
+	// +optional
+	Failure int `json:"failure,omitempty"`
+
+	// +optional
+	Suspend bool `json:"suspend,omitempty"`
+
+	// +optional
+	Conditions []Condition `json:"conditions,omitempty"`
+}
+```
+- `spec.cron` specifies the cron expression for the VM backup schedule
+- `spec.retain` specifies the number of up-to-date VM backups to retain
+- `spec.maxFailure` specifies the max number of consecutive failed backups that could be tolerated, if the number of consecutive failed backups reach this setting, the Harvester controller will suspend the schedule
+- As `spec.resumeRequest` is set, Harvester will try to remove all failed backups and resume the scheduled backup
+- `spec.vmbackup` specifies the source VM to backup
+- `status.vmbackupInfo` records the current backup list, including the status of all the volume backups for the VM
+- `status.failure` number of consecutive failure backups after the last successful backups
+- `status.suspend` presents if the backup schedule is suspended
+- `status.conditions` presents the schedule's status, so far, it's only leveraged to describe the reason for the schedule is suspended
+
+#### ScheduleVMBackup Controller
+
+- ScheduleVMBackup reconcile:
+  - If `.spec.resumeRequest` not set
+    - Create a cronjob with the name `svmbackup-<ScheduleVMBackup UID>` if it's not presented
+    - If the number of the current VM backups goes beyond `spec.retain`
+      - Clean up the oldest VM backup(s)
+      - Delete the outdated Longhorn snapshots to trigger the snapshot purge
```

Comment:
For this one, probably, we can leverage `auto-cleanup-when-delete-backup` setting from https://github.com/longhorn/longhorn/issues/8365.

Reply:
Thanks, @FrankYang0529, I will check it with LH v1.7.0

Reply:
I will also do this in snapshot space management. We can create a PR with LH v1.7.0-rc version and include the setting, so we don't do it twice.

Reply:
https://github.com/harvester/harvester/pull/6236